### PR TITLE
NEW Method selection is now aware of "availability overrides" via Redux reducers

### PR DESCRIPTION
**Pull request by robbieaverill**
This pull request changes a couple of things:

* Move Redux store provider to the boot script
* Add a "withMethodAvailability" HOC which allows Redux reducers from third party modules to provide extra info on whether methods are available in the current context (e.g. if the browser is unsupported)
* Updates enzyme and enzyme-react-wrapper-16 to the latest versions
* Adds a new state for the Verify component where a method is selected but is not available (e.g. if WebAuthn is your default method and you log in using Safari)
* Adds a new "~ExclamationMark~ CircleWarning" component to render this:
image

Todo (separate issue): style the "more options" list when a method is not available

Resolves https://github.com/silverstripe/silverstripe-webauthn-authenticator/issues/13

### CODE REVIEW

File: client/src/components/Verify.scss [hunk 1]
```diff
@@ -2,3 +2,19 @@
...
+
+.mfa-method {
+  text-align: center;
+}
+
+.mfa-method-icon {
+  margin-bottom: $content-spacing;
+
+  &--unavailable {
+    color: $color-blue;
```
Comment (robbieaverill) [inline client/src/components/Verify.scss]: Ah, I think this comes from admin. Will replace.

File: src/RequestHandler/BaseHandlerTrait.php [hunk 2]
```diff
@@ -31,16 +31,16 @@ protected function applyRequirements(bool $frontEndRequirements = true): void
...
         Requirements::add_i18n_javascript('silverstripe/mfa: client/lang');
 
         $suffix = $frontEndRequirements ? '' : '-cms';
         Requirements::javascript("silverstripe/mfa: client/dist/js/bundle{$suffix}.js");
         Requirements::css("silverstripe/mfa: client/dist/styles/bundle{$suffix}.css");
-
-        // Plugin module requirements
-        foreach (MethodRegistry::singleton()->getMethods() as $method) {
-            $method->applyRequirements();
-        }
```
Comment (robbieaverill) [inline src/RequestHandler/BaseHandlerTrait.php]: This was necessary in order to ensure that any method module defined reducers are loaded in time for the redux store in MFA to be created with them

File: src/State/RegisteredMethodDetails.php [hunk 3]
```diff
@@ -24,6 +24,8 @@ public function jsonSerialize()
         return [
             'name' => $this->method->getRegisterHandler()->getName(),
             'urlSegment' => $this->method->getURLSegment(),
+            'isAvailable' => $this->method->isAvailable(),
+            'unavailableMessage' => $this->method->getUnavailableMessage(),
```
Comment (robbieaverill) [inline src/State/RegisteredMethodDetails.php]: I've added these because we need both of them for the verification UI

**Comment by robbieaverill** [comment]
Needs more docs, hold

File: client/src/components/Verify.js [hunk 4]
```diff
@@ -43,6 +55,14 @@ class Verify extends Component {
         : registeredMethods[0]
       );
     }
+
+    // "All methods" is the combination of registered and available methods
+    if (onSetAllMethods) {
+      this.props.onSetAllMethods([
```
Comment (Cheddam) [inline client/src/components/Verify.js]: Would it make more sense to update the store when the data is initially retrieved, rather than during mount?

Comment (robbieaverill) [reply client/src/components/Verify.js]: I think so

**Comment by robbieaverill** [comment]
The SilverStripe 4.1 branches have been deleted, I've removed it from our Travis matrix

**Pull request merged**